### DISCUSSION TITLE
For sites deployed using generate, how to fetch data as part of server start up in Nuxt 3?

### DISCUSSION THREAD

**Discussion by Red-Asuka** [nuxt/nuxt#25501] [Questions] [answered]
There is no introduction to relevant content in the document. I have searched a lot of information and found that someone has the same question as me: https://stackoverflow.com/questions/76488291/how-to-fetch-data-as-part-of-server-start-up-in-nuxt-3

I tried to implement it based on [this answer](https://stackoverflow.com/a/77557211) and it works well under dev, but it doesn't work properly in the production environment. When Nuxt3 generates a static site using "generate", it will statically store the data of "useFetch" as JSON, so that the website can be accessed normally without relying on the backend. I followed this feature when using "useFetch" to fetch data on pages. However, when using "useFetch" to fetch data in plugins, the page will directly call the API.

<img width="929" alt="image" src="https://github.com/nuxt/nuxt/assets/38158783/7a3beffc-a213-44bd-8b23-614326cfd48a">

Code for relevant settings:

```ts
// /server/plugins/fetchVersions.ts

import { useLogger } from '@nuxt/kit'
import type { AllVersionsResponse } from '~/api/downloads'
import { API } from '~/api/downloads'

export default defineNitroPlugin(async () => {
  const storage = useStorage('MQTTX')
  const consola = useLogger()
  const { apiBaseUrl } = useRuntimeConfig().public

  async function fetchVersions() {
    consola.start('Fetching versions...')

    const { data: versions } = await $fetch<ResOptions<AllVersionsResponse>>(`${apiBaseUrl}/${API.ALL_VERSION}`)

    await storage.setItem('versions', versions)
    await storage.setItem('latestVersion', versions[0])

    consola.success(`Versions fetched succesfully!`)
  }
  try {
    fetchVersions()
  }
  catch (e) {
    consola.error('Error fetching versions.')
    process.exit(1)
  }
})
```

```ts
// /server/api/versions.ts

export default cachedEventHandler(async () => {
  return await useStorage('MQTTX').getItem<string[]>('versions')
})
```

```ts
// /plugins/01.injectVersions.ts

export default defineNuxtPlugin(async () => {
  const { data: versions } = await useFetch('/api/versions')

  return {
    provide: {
      versions: versions.value!,
    },
  }
})
```

**Comment by AloisSeckar** [comment, marked as answer]
I have recently read @danielroe `s article that might be an answer to your use case:
https://dev.to/danielroe/using-shared-data-when-generating-pages-45ok

But I haven't tried it personally.

**Reply by danielroe** [reply]
We also just released this feature: https://github.com/nuxt/nuxt/pull/24894.

**Reply by Red-Asuka** [reply]
@AloisSeckar Thank you so much for sharing [nuxt-prepare](https://nuxt-prepare.byjohann.dev/). It has perfectly addressed the issue I was facing. I am sincerely grateful for your assistance.

**Reply by Red-Asuka** [reply]
@danielroe Many thanks to the Nuxt team for your impressive responsiveness. You guys are doing an amazing job!
